cirq-core/cirq/ops/state_preparation_gate.py

```diff
@@ -0,0 +1,125 @@
+
+"""Quantum gates to prepare a given target state."""
+
+from typing import Any, Sequence, Dict, List, Tuple, TYPE_CHECKING
+
+import numpy as np
+
+from cirq import protocols
+from cirq.ops import raw_types
+from cirq.ops.common_channels import ResetChannel
+from cirq.ops.matrix_gates import MatrixGate
+from cirq._compat import proper_repr
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class StatePreparationGate(raw_types.Gate):
+    """A unitary qubit gate which resets all qubits to the |0> state
+    and then prepares the target state."""
+
+    def __init__(self, target_state: np.ndarray, name: str = "StatePreparation") -> None:
+        """Initializes a State Preparation gate.
+
+        Args:
+            target_state: The state vector that this gate should prepare.
+            name: the name of the gate
+
+        Raises:
+            ValueError: if the array is not 1D, or does not have 2**n elements for some integer n.
+        """
+        if len(target_state.shape) != 1:
+            raise ValueError('`target_state` must be a 1d numpy array.')
+
+        n = int(np.round(np.log2(target_state.shape[0] or 1)))
+        if 2 ** n != target_state.shape[0]:
+            raise ValueError(f'Matrix width ({target_state.shape[0]}) is not a power of 2')
+
+        self._state = target_state.astype(np.complex128) / np.linalg.norm(target_state)
+        self._num_qubits = n
+        self._name = name
+        self._qid_shape = (2,) * n
+
+    @staticmethod
+    def _has_unitary_() -> bool:
+        """Checks and returns if the gate has a unitary representation.
+        It doesn't, since the resetting of the channels is a non-unitary operations,
+        it involves measurement."""
+        return False
+
+    def _json_dict_(self) -> Dict[str, Any]:
+        """Converts the gate object into a serializable dictionary"""
+        return {
+            'cirq_type': self.__class__.__name__,
+            'target_state': self._state.tolist(),
+        }
+
+    @classmethod
+    def _from_json_dict_(cls, target_state, **kwargs):
+        """Recreates the gate object from it's serialized form
+
+        Args:
+            target_state: the state to prepare using this gate
+            kwargs: other keyword arguments, ignored
+        """
+        return cls(target_state=np.array(target_state))
+
+    def _num_qubits_(self):
+        return self._num_qubits
+
+    def _qid_shape_(self) -> Tuple[int, ...]:
+        return self._qid_shape
+
+    def _circuit_diagram_info_(
+        self, _args: 'cirq.CircuitDiagramInfoArgs'
+    ) -> 'cirq.CircuitDiagramInfo':
+        """Returns the information required to draw out the circuit diagram for this gate."""
+        symbols = (
+            [self._name]
+            if self._num_qubits == 1
+            else [f'{self._name}[{i+1}]' for i in range(0, self._num_qubits)]
+        )
+        return protocols.CircuitDiagramInfo(wire_symbols=symbols)
+
+    def _get_unitary_transform(self):
```

Comment:
Instead of trying to find a unitary that takes `|0>` to `|\Psi>`, I'd propose that we should just define state preparation as a channel with kraus operators `|\Psi><i|` for all `|i>`.  The operators satisfy the completeness relation `Sum(E^ E) = I`. 

This should be enough to define the behaviour of the gate during simulation. For specific target gatesets, efficient unitary constructions can be provided by the compilers consuming this gate.

Reply:
Thanks a lot, this fixed all the issues I was having in tests with complex numbers.

Reply:
Great. Can you please update the title and description of the PR to reflect the new approach?

Reply:
Done.